File: src/main/java/com/everyonewaiter/infrastructure/image/OracleObjectStorageClient.java
package com.everyonewaiter.infrastructure.image;

import com.everyonewaiter.domain.image.service.ImageClient;
import com.everyonewaiter.domain.shared.BusinessException;
import com.everyonewaiter.domain.shared.ErrorCode;
import com.oracle.bmc.ConfigFileReader;
import com.oracle.bmc.Region;
import com.oracle.bmc.auth.ConfigFileAuthenticationDetailsProvider;
import com.oracle.bmc.http.client.StandardClientProperties;
import com.oracle.bmc.objectstorage.ObjectStorageClient;
import com.oracle.bmc.objectstorage.requests.DeleteObjectRequest;
import com.oracle.bmc.objectstorage.requests.PutObjectRequest;
import com.oracle.bmc.objectstorage.transfer.UploadConfiguration;
import com.oracle.bmc.objectstorage.transfer.UploadManager;
import java.io.File;
import java.io.IOException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
class OracleObjectStorageClient implements ImageClient {

  private final OracleObjectStorageProperties properties;

  @Override
  public void upload(File imageFile, String imageName, String contentType) {
    try (ObjectStorageClient client = initializeClient()) {
      UploadConfiguration configuration = UploadConfiguration.builder()
          .allowMultipartUploads(true)
          .allowParallelUploads(true)
          .build();
      UploadManager uploadManager = new UploadManager(client, configuration);

      UploadManager.UploadRequest request = UploadManager.UploadRequest.builder(imageFile)
          .allowOverwrite(true)
          .build(
              PutObjectRequest.builder()
                  .namespaceName(properties.getNamespace())
                  .bucketName(properties.getBucketName())
                  .objectName(imageName)
                  .contentType(contentType)
                  .contentLength(imageFile.length())
                  .cacheControl("max-age=86400")
                  .build()
          );

      uploadManager.upload(request);
    } catch (Exception exception) {
      throw new BusinessException(ErrorCode.FAILED_UPLOAD_IMAGE);
    }
  }

  @Override
  public void delete(String imageName) {
    try (ObjectStorageClient client = initializeClient()) {
      DeleteObjectRequest request = DeleteObjectRequest.builder()
          .namespaceName(properties.getNamespace())
          .bucketName(properties.getBucketName())
          .objectName(imageName)
          .build();
      client.deleteObject(request);
    } catch (Exception e) {
      throw new BusinessException(ErrorCode.FAILED_DELETE_IMAGE);
    }
  }

  private ObjectStorageClient initializeClient() throws IOException {
    ConfigFileReader.ConfigFile configFile = ConfigFileReader.parseDefault();
    return ObjectStorageClient.builder()
        .region(Region.AP_SEOUL_1)
        .clientConfigurator(builder ->
            builder.property(StandardClientProperties.BUFFER_REQUEST, false)
        )
        .build(new ConfigFileAuthenticationDetailsProvider(configFile));
  }

}

File: src/main/kotlin/com/everyonewaiter/infrastructure/image/OracleObjectStorageClient.kt
package com.everyonewaiter.infrastructure.image

import com.everyonewaiter.common.file.extension.deleteWithLog
import com.everyonewaiter.common.file.extension.extensionOrThrow
import com.everyonewaiter.common.file.extension.hasExtension
import com.everyonewaiter.common.file.extension.toFile
import com.everyonewaiter.global.exception.BusinessException
import com.everyonewaiter.global.exception.ErrorCode
import com.everyonewaiter.global.extension.checkOrThrow
import com.oracle.bmc.objectstorage.ObjectStorageClient
import com.oracle.bmc.objectstorage.requests.PutObjectRequest
import com.oracle.bmc.objectstorage.transfer.UploadConfiguration
import com.oracle.bmc.objectstorage.transfer.UploadManager
import com.oracle.bmc.objectstorage.transfer.UploadManager.UploadRequest
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import org.springframework.web.multipart.MultipartFile

@Component
class OracleObjectStorageClient(
    @Value("\${oracle.object.storage.namespace}") val namespace: String,
    @Value("\${oracle.object.storage.bucket-name}") val bucketName: String,
    private val objectStorageClient: ObjectStorageClient,
) : ImageClient {
    override fun upload(file: MultipartFile): String {
        checkOrThrow(file.originalFilename != null, ErrorCode.NOT_FOUND_FILENAME)
        checkOrThrow(file.hasExtension, ErrorCode.NOT_FOUND_EXTENSION)
        checkOrThrow(file.contentType != null, ErrorCode.NOT_FOUND_CONTENT_TYPE)
        val convertedFile = file.toFile
        try {
            objectStorageClient.use { client ->
                val request = UploadRequest
                    .builder(convertedFile)
                    .allowOverwrite(true)
                    .build(
                        PutObjectRequest
                            .builder()
                            .namespaceName(namespace)
                            .bucketName(bucketName)
                            .objectName(file.originalFilename)
                            .contentType(file.contentType)
                            .contentLength(convertedFile.length())
                            .contentDisposition("attachment; filename=${file.name}.${file.extensionOrThrow}")
                            .cacheControl("max-age=86400")
                            .build(),
                    )
                val configuration = UploadConfiguration
                    .builder()
                    .allowMultipartUploads(true)
                    .allowParallelUploads(true)
                    .build()
                UploadManager(client, configuration).upload(request)
                return file.originalFilename!!
            }
        } catch (exception: Exception) {
            throw BusinessException(ErrorCode.FAILED_UPLOAD_IMAGE)
        } finally {
            convertedFile.deleteWithLog()
        }
    }
}
